src/CredentialSource/ExecutableSource.php

```diff
@@ -0,0 +1,217 @@
+ *   "id_token": "HEADER.PAYLOAD.SIGNATURE",
+ *   "expiration_time": 1620433341
+ * }
+ *
+ * SAML2 response sample:
+ * {
+ *   "version": 1,
+ *   "success": true,
+ *   "token_type": "urn:ietf:params:oauth:token-type:saml2",
+ *   "saml_response": "...",
+ *   "expiration_time": 1620433341
+ * }
+ *
+ * Error response sample:
+ * {
+ *   "version": 1,
+ *   "success": false,
+ *   "code": "401",
+ *   "message": "Error message."
+ * }
+ * </pre>
+ *
+ * The "expiration_time" field in the JSON response is only required for successful
+ * responses when an output file was specified in the credential configuration
+ *
+ * The auth libraries will populate certain environment variables that will be accessible by the
+ * executable, such as: GOOGLE_EXTERNAL_ACCOUNT_AUDIENCE, GOOGLE_EXTERNAL_ACCOUNT_TOKEN_TYPE,
+ * GOOGLE_EXTERNAL_ACCOUNT_INTERACTIVE, GOOGLE_EXTERNAL_ACCOUNT_IMPERSONATED_EMAIL, and
+ * GOOGLE_EXTERNAL_ACCOUNT_OUTPUT_FILE.
+ */
+class ExecutableSource implements ExternalAccountCredentialSourceInterface
+{
+    private const GOOGLE_EXTERNAL_ACCOUNT_ALLOW_EXECUTABLES = 'GOOGLE_EXTERNAL_ACCOUNT_ALLOW_EXECUTABLES';
+    private const SAML_SUBJECT_TOKEN_TYPE = 'urn:ietf:params:oauth:token-type:saml2';
+    private const OIDC_SUBJECT_TOKEN_TYPE1 = 'urn:ietf:params:oauth:token-type:id_token';
+    private const OIDC_SUBJECT_TOKEN_TYPE2 = 'urn:ietf:params:oauth:token-type:jwt';
+
+    private string $command;
+    private ExecutableHandler $executableHandler;
+    private ?string $outputFile;
+
+    /**
+     * @param string $command    The string command to run to get the subject token.
+     * @param string $outputFile
+     */
+    public function __construct(
+        string $command,
+        ?string $outputFile,
+        ExecutableHandler $executableHandler = null,
+    ) {
+        $this->command = $command;
+        $this->outputFile = $outputFile;
+        $this->executableHandler = $executableHandler ?: new ExecutableHandler();
+    }
+
+    /**
+     * @param callable $httpHandler unused.
+     */
+    public function fetchSubjectToken(callable $httpHandler = null): string
+    {
+        // Check if the executable is allowed to run.
+        if (getenv(self::GOOGLE_EXTERNAL_ACCOUNT_ALLOW_EXECUTABLES) !== '1') {
+            throw new RuntimeException(
+                'Pluggable Auth executables need to be explicitly allowed to run by '
+                . 'setting the GOOGLE_EXTERNAL_ACCOUNT_ALLOW_EXECUTABLES environment '
+                . 'Variable to 1.'
+            );
+        }
+
+        if (
+            $this->outputFile
+            && file_exists($this->outputFile)
+            && !empty($outputFileContents = file_get_contents($this->outputFile))
+        ) {
+            $json = json_decode($outputFileContents, true);
+            if (isset($json['expiration_time']) && time() < $json['expiration_time']) {
+                return $this->parseTokenFromResponse($outputFileContents);
```

Comment:
nit: If there is an error here, the messages won't specify that it comes from reading an output file vs running the executable directly right? Might be helpful to differentiate that for the user.

Reply:
Also, if the file response is valid Json but unsuccessful what happens here? In that case we want to re-run the executable.

Reply:
> If there is an error here, the messages won't specify that it comes from reading an output file vs running the executable directly right?

That's correct. It should be easy to add that

> if the file response is valid Json but unsuccessful what happens here?

Good question, I believe it throws an error same as if it's the response. Are you suggesting that any errors in the file with `success: false` should silently retry the executable?

Is there a good document that lists the expected logic for the `outputFile`, because it's not very intuitive. I followed the [NodeJS implementation](https://github.com/googleapis/google-auth-library-nodejs/blob/main/src/auth/pluggable-auth-client.ts), although I may have missed some things.

Reply:
go/pluggable-auth-design but it doesn't look like that has a great flow diagram. But yes, the expected behavior is that if the output file is parseable (I.e. good JSON) but success:false, then we would re-run the executable instead of just returning the error again

Reply:
I have manually run the tests in go/pluggable-auth-sdk-testing and they are now passing